_posts/2016-09-19-variance-phantom.md

```diff
@@ -0,0 +1,377 @@
+[the missing fourth variance]({% post_url 2016-02-04-variance-and-functors %}#one-more-thing).
+I like to write it as I did in
+[“The missing diamond of Scala variance”](https://failex.blogspot.com/2016/09/the-missing-diamond-of-scala-variance.html):
+
+```scala
+sealed abstract class MyModel[👻P, I, -T, +V]
+```
+
+Second, and more practically, it illuminates the role of variance in
+pattern matching in a way that can be difficult to see with that
+confusing data in the way.
+
+## It can be covariant
+
+The rule for a type parameter being parametrically covariant says we
+have to look at all the positions in the data where the type parameter
+occurs; if every one of them is a covariant position, then the type
+parameter may be marked covariant. Consider a simplified version of
+`MyModel`.
+
+```scala
+sealed abstract class Gimme[P]
+case object AStr extends Gimme[String]
+case object AnInt extends Gimme[Int]
+```
+
+The type parameter `P` appears in no positions, so it vacuously
+satisfies the requirement “every occurrence is in covariant position”.
+
+So let us mark `P` covariant.
+
+```scala
+sealed abstract class Gimme[+P]
+// otherwise the same
+```
+
+## It can be contravariant
+
+The rule for contravariance is also based on the occurrences of the
+type parameter: if every occurrence is in contravariant position, then
+the type parameter may be contravariant.
+
+This rule seems to be contradict the rule for covariance, except that
+all “every” statements are always true when the set under
+consideration is empty.
+
+1. Set **S** is empty.
+2. Every element of set **S** is a dog.
+3. No element of set **S** is a dog.
+
+2 and 3 can be true at the same time, but only if 1 is true, too. So
+let us mark `P` contravariant, in a renamed ADT.
+
+```scala
+sealed abstract class Gotme[-P]
+case object UnStr extends Gotme[String]
+case object UnInt extends Gotme[Int]
+```
+
+## The usual relationships
+
+Since you can choose any variance for phantom parameters, the
+important question is: what kind of type relationships should
+exist within my ADT?
+
+At first, this seems to be merely a question of how values of `Gimme`
+and `Gotme` types ought to widen.
+
+1. Every `Gimme[Cat]` is a `Gimme[Animal]`, and
+2. every `Gotme[Animal]` is a `Gotme[Cat]`. Moreover,
+3. every `Gimme[Nothing]` is a `Gimme[T]` no matter what `T` is, and
+4. every `Gotme[Any]` is a `Gotme[T]` no matter what `T` is.
+
+Obviously, if neither of these behaviors—the 1/3 nor the 2/4—is
+desirable, you shouldn’t use variance. In my experience, this is the
+case for most phantom types. If one is desirable, then it may be fine,
+but there’s more to consider.
+
+## Extracting the covariant
+
+Pattern-matching on the covariant `Gimme` reveals fully safe type
+information. Unlike `ClassTag` and `TypeTag`, which are egregiously
+broken for this use case, this method of carrying type information
+forward into runtime is closed and Scalazzi-safe.
```

Comment:
@larsrh I mean, matching in general isn't safe, because of `scalac` bugs, or because of consequences of upstream casts (e.g. "the singleton instance trick).  Do you mean something more theoretical?